Tech debt: Idempotency Thread safety E2E test is flaky

### Why is this needed?

Ensure E2E pipeline works successfully.

---

```python
FAILED tests/e2e/idempotency/test_idempotency_dynamodb.py::test_idempotent_function_thread_safety@idempotency - assert {'3', 'l', '1', 'x', '_', '-', '8', '0', 't', '}', 'u', 'a', 'H', 'i', 'o', 'h', 'e', 'p', 'E', 'd', 'S', 'P', '6', ',', '[', '{', 'c', 'I', 'n', '7', ' ', 'D', 'N', ']', 'm', 's', 'r', 'T', 'F', '2', '5', ':', '.', '"'} == {'3', 'l', '1', 'x', '_', '-', '8', '0', 't', '}', 'u', 'a', 'H', '4', 'i', 'o', 'h', 'e', 'p', 'E', 'd', 'S', 'P', '6', ',', '[', '9', '{', 'c', 'I', 'n', '7', ' ', 'D', 'N', ']', 'm', 's', 'r', 'T', 'F', '2', '5', ':', '.', '"'}
  Extra items in the right set:
  '9'
  '4'
  Full diff:
    {
     ' ',
     '"',
     ',',
     '-',
     '.',
     '0',
     '1',
     '2',
     '3',
  -  '4',
     '5',
     '6',
     '7',
     '8',
  -  '9',
     ':',
     'D',
     'E',
     'F',
     'H',
     'I',
     'N',
     'P',
     'S',
     'T',
     '[',
     ']',
     '_',
     'a',
     'c',
     'd',
     'e',
     'h',
     'i',
     'l',
     'm',
     'n',
     'o',
     'p',
     'r',
     's',
     't',
     'u',
     'x',
     '{',
     '}',
    }
============ 1 failed, 50 passed, 10 warnings in 233.35s (0:03:53) =============
```

### Which area does this relate to?

Tests

### Suggestion

_No response_

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

--- a/aws_lambda_powertools/utilities/idempotency/persistence/base.py
+++ b/aws_lambda_powertools/utilities/idempotency/persistence/base.py
@@ -196,6 +196,8 @@ class BasePersistenceLayer(ABC):
     def is_missing_idempotency_key(data) -> bool:
         if type(data).__name__ in ("tuple", "list", "dict"):
             return all(x is None for x in data)
+        elif isinstance(data, (int, float, bool)):
+            return False
         return not data
 
     def _get_hashed_payload(self, data: Dict[str, Any]) -> str:

--- a/aws_lambda_powertools/utilities/idempotency/persistence/base.py
+++ b/aws_lambda_powertools/utilities/idempotency/persistence/base.py
@@ -194,7 +194,7 @@ class BasePersistenceLayer(ABC):
 
     @staticmethod
     def is_missing_idempotency_key(data) -> bool:
-        if type(data).__name__ in ("tuple", "list", "dict"):
+        if isinstance(data, (tuple, list, dict)):
             return all(x is None for x in data)
         elif isinstance(data, (int, float, bool)):
             return False

--- a/tests/e2e/event_handler/infrastructure.py
+++ b/tests/e2e/event_handler/infrastructure.py
@@ -73,7 +73,14 @@ class EventHandlerStack(BaseInfrastructure):
         CfnOutput(self.stack, "APIGatewayHTTPUrl", value=(apigw.url or ""))
 
     def _create_api_gateway_rest(self, function: Function):
-        apigw = apigwv1.RestApi(self.stack, "APIGatewayRest", deploy_options=apigwv1.StageOptions(stage_name="dev"))
+        apigw = apigwv1.RestApi(
+            self.stack,
+            "APIGatewayRest",
+            deploy_options=apigwv1.StageOptions(stage_name="dev"),
+            # disables creation of a role that is not destroyed due to a breaking change in CDK
+            # https://github.com/aws/aws-cdk/issues/22020
+            cloud_watch_role=False,
+        )
 
         todos = apigw.root.add_resource("todos")
         todos.add_method("POST", apigwv1.LambdaIntegration(function, proxy=True))

--- a/tests/functional/idempotency/test_idempotency.py
+++ b/tests/functional/idempotency/test_idempotency.py
@@ -952,11 +952,6 @@ def test_is_missing_idempotency_key():
     assert BasePersistenceLayer.is_missing_idempotency_key({})
     # GIVEN an empty str THEN is_missing_idempotency_key is True
     assert BasePersistenceLayer.is_missing_idempotency_key("")
-    # GIVEN False THEN is_missing_idempotency_key is True
-    assert BasePersistenceLayer.is_missing_idempotency_key(False)
-    # GIVEN number 0 THEN is_missing_idempotency_key is True
-    assert BasePersistenceLayer.is_missing_idempotency_key(0)
-
     # GIVEN None THEN is_missing_idempotency_key is True
     assert BasePersistenceLayer.is_missing_idempotency_key(None)
     # GIVEN a list of Nones THEN is_missing_idempotency_key is True
@@ -966,15 +961,23 @@ def test_is_missing_idempotency_key():
     # GIVEN a dict of Nones THEN is_missing_idempotency_key is True
     assert BasePersistenceLayer.is_missing_idempotency_key({None: None})
 
+    # GIVEN True THEN is_missing_idempotency_key is False
+    assert BasePersistenceLayer.is_missing_idempotency_key(True) is False
+    # GIVEN False THEN is_missing_idempotency_key is False
+    assert BasePersistenceLayer.is_missing_idempotency_key(False) is False
+    # GIVEN number 0 THEN is_missing_idempotency_key is False
+    assert BasePersistenceLayer.is_missing_idempotency_key(0) is False
+    # GIVEN number 0.0 THEN is_missing_idempotency_key is False
+    assert BasePersistenceLayer.is_missing_idempotency_key(0.0) is False
     # GIVEN a str THEN is_missing_idempotency_key is False
     assert BasePersistenceLayer.is_missing_idempotency_key("Value") is False
     # GIVEN str "False" THEN is_missing_idempotency_key is False
     assert BasePersistenceLayer.is_missing_idempotency_key("False") is False
-    # GIVEN an number THEN is_missing_idempotency_key is False
+    # GIVEN a number THEN is_missing_idempotency_key is False
     assert BasePersistenceLayer.is_missing_idempotency_key(1000) is False
     # GIVEN a float THEN is_missing_idempotency_key is False
     assert BasePersistenceLayer.is_missing_idempotency_key(10.01) is False
-    # GIVEN a list of all not None THEN is_missing_idempotency_key is False
+    # GIVEN a list with some items THEN is_missing_idempotency_key is False
     assert BasePersistenceLayer.is_missing_idempotency_key([None, "Value"]) is False
 
 

